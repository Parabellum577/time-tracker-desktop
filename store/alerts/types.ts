import { ActionType } from 'typesafe-actions'

import * as actions from './actions'
import { IError } from '@services/types'

export type AlertsAction = ActionType<typeof actions>

export interface IAlert {
  alertType: 'success' | 'error' | 'warning' | 'coldstart'
  alertMessage: string | IError
}

export interface IAlertsState {
  alertsArray: IAlert[]
}
