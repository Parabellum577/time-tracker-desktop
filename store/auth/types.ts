import { ActionType } from 'typesafe-actions'
import { DeepReadonly } from 'utility-types'

import * as actions from './actions'
import { IError } from '@types'

export type AuthAction = ActionType<typeof actions>

export type IAuthState = DeepReadonly<{
  readonly authorized: boolean
  readonly error?: IError | null
  readonly loading: boolean
  readonly isLoadingData: boolean
}>
