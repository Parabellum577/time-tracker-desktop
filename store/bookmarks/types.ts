import { IFullIssue } from '@types'

export interface IBookmarks {
  extendedBookmarks: IFullIssue[]
}
