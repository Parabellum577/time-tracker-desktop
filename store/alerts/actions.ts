import { createAsyncAction } from 'typesafe-actions'
import { IError } from '@types'
import { IAlert } from './types'

export const showAlertMessage = createAsyncAction(
  'notifications/SHOW_ALERT_MESSAGE_REQUEST',
  'notifications/SHOW_ALERT_MESSAGE_SUCCESS',
  'notifications/SHOW_ALERT_MESSAGE_FAILURE',
)<IAlert, IAlert, IError>()

export const clearAlertMessage = createAsyncAction(
  'notifications/CLEAR_ALERT_MESSAGE_REQUEST',
  'notifications/CLEAR_ALERT_MESSAGE_SUCCESS',
  'notifications/CLEAR_ALERT_MESSAGE_FAILURE',
)<void, void, IError>()
