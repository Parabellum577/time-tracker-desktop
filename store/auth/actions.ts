import { createAction, createAsyncAction } from 'typesafe-actions'
import { IError } from '@types'
import { ILoginUserPayload, ILoginOTSPayload } from '@services/api-types'

export const loginUser = createAsyncAction('auth/LOGIN_REQUEST', 'auth/LOGIN_SUCCESS', 'auth/LOGIN_FAILURE')<
  ILoginUserPayload,
  void,
  IError
>()

export const loginUserViaOAuth = createAsyncAction(
  'auth/LOGIN_VIA_OAUTH_REQUEST',
  'auth/LOGIN_VIA_OAUTH_SUCCESS',
  'auth/LOGIN_VIA_OAUTH_FAILURE',
)<void, void, IError>()

export const otsLogin = createAsyncAction('auth/OTS_LOGIN_REQUEST', 'auth/OTS_LOGIN_SUCCESS', 'auth/OTS_LOGIN_FAILURE')<
  ILoginOTSPayload,
  ILoginOTSPayload,
  IError
>()

export const loadingDataAfterLogin = createAction('auth/LOADING_DATA_AFTER_LOGIN', resolve => (isLoading: boolean) =>
  resolve(isLoading),
)

export const signOut = createAction('auth/SIGN_OUT', resolve => () => resolve())
