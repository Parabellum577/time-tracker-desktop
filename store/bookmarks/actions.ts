import { createAsyncAction, createAction } from 'typesafe-actions'
import { IError, IFullIssue } from '@types'

export const getBookmarks = createAsyncAction(
  'bookmarks/GET_BOOKMARKS_REQUEST',
  'bookmarks/GET_BOOKMARKS_SUCCESS',
  'bookmarks/GET_BOOKMARKS_FAILURE',
)<void, IFullIssue[], IError>()

export const handleBookmark = createAsyncAction(
  'bookmarks/HANDLE_BOOKMARK_REQUEST',
  'bookmarks/HANDLE_BOOKMARK_SUCCESS',
  'bookmarks/HANDLE_BOOKMARK_FAILURE',
)<IFullIssue, IFullIssue, IError>()

export const clearBookmarks = createAction('plannings/CLEAR_BOOKMARKS', action => () => action())
