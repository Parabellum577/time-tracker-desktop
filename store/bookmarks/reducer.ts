import { getType, ActionType } from 'typesafe-actions'

import * as actions from './actions'
import { IBookmarks } from './types'

const initialState: IBookmarks = {
  extendedBookmarks: [],
}

export default (state: IBookmarks = initialState, action: ActionType<typeof actions>): IBookmarks => {
  switch (action.type) {
    case getType(actions.getBookmarks.success):
      return { ...state, extendedBookmarks: action.payload }
    case getType(actions.handleBookmark.success):
      const isRemove = state.extendedBookmarks.some(bookmark => bookmark.ID === action.payload.ID)
      if (isRemove) {
        return {
          ...state,
          extendedBookmarks: state.extendedBookmarks.filter(bookmark => bookmark.ID !== action.payload.ID),
        }
      } else {
        return { ...state, extendedBookmarks: [...state.extendedBookmarks, action.payload] }
      }
    case getType(actions.clearBookmarks):
      return { ...initialState }
    default:
      return state
  }
}
