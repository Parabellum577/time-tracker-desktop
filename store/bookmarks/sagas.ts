import { call, take, race, put, select } from 'redux-saga/effects'
import { SagaIterator } from 'redux-saga'
import { getType } from 'typesafe-actions'

import * as loginActions from '../auth/actions'
import * as alertActions from '../alerts/actions'
import * as actions from './actions'
import api from '@api'
import { IBasicIssue, IFullIssue } from '@types'
import extendIssues from '@services/extendIssues'
import { IRootState } from '@store/rootReducer'

export function* getBookmarksSaga(): SagaIterator {
  while (true) {
    yield race({
      getBookmarksAction: take(getType(actions.getBookmarks.request)),
      loginUserAction: take(getType(loginActions.loginUser.success)),
    })
    try {
      const basicBookmarks: IBasicIssue[] = yield call(api.tracker.GetBookmarks)
      const bookmarks = yield call(extendIssues, basicBookmarks)
      yield put(actions.getBookmarks.success(bookmarks))
    } catch (error) {
      console.error(error)
    }
  }
}

export const bookmarksArrayFromStore = (state: IRootState) => state.bookmarks.extendedBookmarks

export function* handleBookmarkSaga(): SagaIterator {
  while (true) {
    const { payload } = yield take(getType(actions.handleBookmark.request))
    try {
      const bookmarks: IFullIssue[] = yield select(bookmarksArrayFromStore)
      const isRemove = bookmarks.some(bookmark => bookmark.ID === payload.ID)
      const bookmarkPayload = {
        TrackerID: payload.TrackerID,
        ProjectID: payload.ProjectID,
        IssueID: payload.ID,
      }
      yield put(actions.handleBookmark.success(payload))
      if (isRemove) {
        yield call(api.tracker.RemoveBookmark, bookmarkPayload)
        yield put(
          alertActions.showAlertMessage.request({
            alertType: 'success',
            alertMessage: 'Removed from Favorites',
          }),
        )
      } else {
        yield call(api.tracker.CreateBookmark, bookmarkPayload)
        yield put(
          alertActions.showAlertMessage.request({
            alertType: 'success',
            alertMessage: 'Added to Favorites',
          }),
        )
      }
    } catch (error) {
      console.error('handleBookmarkSaga: ', error)
    }
  }
}
