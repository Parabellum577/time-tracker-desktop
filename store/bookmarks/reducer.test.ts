import { ActionType, getType } from 'typesafe-actions'

import * as actions from './actions'
import { IBookmarks } from './types'
import bookmarks from './reducer'
import { SIMPLE_ISSUE } from '@test-helpers/sample'

describe('[Store] Bookmarks Reducer', () => {
  it('[Success] [Snapshot] Should handle getBookmarks', () => {
    const getBookmarksAction: ActionType<typeof actions> = {
      type: getType(actions.getBookmarks.success),
      payload: [{ ...SIMPLE_ISSUE }, { ...SIMPLE_ISSUE }],
    }

    const currentState: IBookmarks = {
      extendedBookmarks: [{ ...SIMPLE_ISSUE }, { ...SIMPLE_ISSUE }],
    }

    expect(bookmarks(currentState, getBookmarksAction)).toMatchSnapshot()
  })

  it('[Failure] Should handle getBookmarks', () => {
    const getBookmarksAction: ActionType<typeof actions> = {
      type: getType(actions.getBookmarks.failure),
      payload: {
        code: 666,
        message: 'SOME_ERROR',
      },
    }

    const currentState: IBookmarks = {
      extendedBookmarks: [],
    }

    expect(bookmarks(currentState, getBookmarksAction)).toMatchSnapshot()
  })
})
