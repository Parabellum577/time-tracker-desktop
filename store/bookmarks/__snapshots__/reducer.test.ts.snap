// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Store] Bookmarks Reducer [Failure] Should handle getBookmarks 1`] = `
Object {
  "extendedBookmarks": Array [],
}
`;

exports[`[Store] Bookmarks Reducer [Success] [Snapshot] Should handle getBookmarks 1`] = `
Object {
  "extendedBookmarks": Array [
    Object {
      "Activity": 0,
      "Done": 0,
      "DueDate": 0,
      "Estimate": 0,
      "ID": "string",
      "IsAssigned": true,
      "Priority": Object {
        "ID": 0,
        "Name": "string",
      },
      "ProjectID": "string",
      "Spent": 0,
      "Status": Object {
        "ID": 0,
        "Name": "string",
      },
      "Title": "string",
      "TrackerID": 0,
      "Type": Object {
        "ID": 0,
        "Name": "string",
      },
      "URL": "string",
    },
    Object {
      "Activity": 0,
      "Done": 0,
      "DueDate": 0,
      "Estimate": 0,
      "ID": "string",
      "IsAssigned": true,
      "Priority": Object {
        "ID": 0,
        "Name": "string",
      },
      "ProjectID": "string",
      "Spent": 0,
      "Status": Object {
        "ID": 0,
        "Name": "string",
      },
      "Title": "string",
      "TrackerID": 0,
      "Type": Object {
        "ID": 0,
        "Name": "string",
      },
      "URL": "string",
    },
  ],
}
`;
